src/commonMain/kotlin/fr/acinq/lightning/utils/logger.kt

```diff
@@ -1,10 +1,96 @@
+import fr.acinq.lightning.payment.PaymentPart
+import fr.acinq.lightning.wire.HasChannelId
+import fr.acinq.lightning.wire.HasTemporaryChannelId
+import fr.acinq.lightning.wire.LightningMessage
 import org.kodein.log.Logger
+import org.kodein.log.Logger.Level.*
 
 /**
  * This should be used more largely once https://kotlinlang.org/docs/whatsnew1620.html#prototype-of-context-receivers-for-kotlin-jvm is stable
  */
 interface LoggingContext {
-    val logger: Logger
+    val logger: MDCLogger
+}
+
+/**
+ * A simpler wrapper on top of [Logger] with better MDC support.
+ */
+data class MDCLogger(val logger: Logger, val staticMdc: Map<String, Any> = emptyMap()) {
+    inline fun debug(mdc: Map<String, Any> = emptyMap(), msgCreator: () -> String) {
+        logger.log(level = DEBUG, msgCreator = msgCreator, meta = staticMdc + mdc)
+    }
+
+    inline fun info(mdc: Map<String, Any> = emptyMap(), msgCreator: () -> String) {
+        logger.log(level = INFO, msgCreator = msgCreator, meta = staticMdc + mdc)
+    }
+
+    inline fun warning(ex: Throwable? = null, mdc: Map<String, Any> = emptyMap(), msgCreator: () -> String) {
+        logger.log(level = WARNING, error = ex, msgCreator = msgCreator, meta = staticMdc + mdc)
+    }
+
+    inline fun error(ex: Throwable? = null, mdc: Map<String, Any> = emptyMap(), msgCreator: () -> String) {
+        logger.log(level = ERROR, error = ex, msgCreator = msgCreator, meta = staticMdc + mdc)
+    }
+}
+
+suspend fun <T> MDCLogger.withMDC(mdc: Map<String, Any>, f: suspend (MDCLogger) -> T): T {
+    val logger = this.copy(staticMdc = this.staticMdc + mdc)
+    return f(logger)
+}
+
+/**
+ * Utility functions to build MDC for various objects without polluting main classes
+ */
+
+fun PaymentPart.mdc(): Map<String, Any> = mapOf(
+    "paymentHash" to paymentHash,
+    "amount" to amount,
+    "totalAmount" to totalAmount
+)
+
+fun SendPayment.mdc(): Map<String, Any> = mapOf(
+    "paymentId" to paymentId,
+    "paymentHash" to paymentHash,
+    "amount" to amount,
+    "recipient" to recipient
+)
+
+fun OutgoingPayment.mdc(): Map<String, Any> = mapOf(
+    "paymentId" to id,
+    "paymentHash" to paymentHash,
+    "amount" to amount,
+    "recipient" to recipient
+)
+
+fun ChannelState.mdc(): Map<String, Any> {
+    val state = this
+    return buildMap {
+        state::class.simpleName?.let { put("state", it) }
+        when (state) {
+            is WaitForOpenChannel -> put("temporaryChannelId", state.temporaryChannelId)
+            is WaitForAcceptChannel -> put("temporaryChannelId", state.temporaryChannelId)
+            is WaitForFundingCreated -> put("channelId", state.channelId)
+            is WaitForFundingSigned -> put("channelId", state.channelId)
+            is ChannelStateWithCommitments -> put("channelId", state.channelId)
+            is Offline -> put("channelId", state.state.channelId)
+            is Syncing -> put("channelId", state.state.channelId)
+            else -> {}
```

Comment:
Shouldn't we have exhaustive pattern matching here?

Reply:
Some states like `Aborted` don't include a channel id.

Reply:
Yes but that's not really the point, you can have exhaustive pattern matching and not do anything for the states that don't include a `channelId`. That makes sure we're doing the right thing for all states, whereas here a state that does have a `channelId` may be incorrectly included in the `else` branch. But it's nit, it looks like we do list all the states that have a `channelId`, so it's not very important.